src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+
+        // update each quorum's aggregate pubkey
+        _processQuorumApkUpdate(quorumNumbers, pubkey.negate());
+        // update the global aggregate pubkey
+        _processGlobalApkUpdate(pubkey.negate());
+        // emit event so offchain actors can update their state
+        emit PubkeyRemoved(operator, pubkey);
+        return pubkeyHash;
+    }
+
+    /// @notice Returns the current APK for the provided `quorumNumber `
+    function getApkForQuorum(uint8 quorumNumber) external view returns(BN254.G1Point memory) {
+        return quorumApk[quorumNumber];
+    }
+
+    /// @notice Returns the `ApkUpdate` struct at `index` in the list of APK updates for the `quorumNumber`
+    function getApkUpdateForQuorumByIndex(uint8 quorumNumber, uint256 index) external view returns (ApkUpdate memory){
+        return quorumApkUpdates[quorumNumber][index];
+    }
+
+    /**
+     * @notice get hash of the apk of `quorumNumber` at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     * @param quorumNumber is the quorum whose ApkHash is being retrieved
+     * @param blockNumber is the number of the block for which the latest ApkHash muust be retrieved
+     * @param index is the provided witness of the onchain index calculated offchain
+     */
+    function getApkHashForQuorumAtBlockNumberFromIndex(uint8 quorumNumber, uint32 blockNumber, uint256 index) external returns (bytes32){
+        ApkUpdate memory quorumApkUpdate = quorumApkUpdates[quorumNumber][index];
+        _validateApkHashForQuorumAtBlockNumber(quorumApkUpdate, blockNumber);
+        return quorumApkUpdate.apkHash;
+    }
+
+	/**
+     * @notice get hash of the apk among all quourums at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     */
+    function getGlobalApkHashAtBlockNumberFromIndex(uint32 blockNumber, uint256 index) external view returns (bytes32){
+        ApkUpdate memory globalApkUpdate = globalApkUpdates[index];
+        _validateApkHashForQuorumAtBlockNumber(globalApkUpdate, blockNumber);
+        return globalApkUpdate.apkHash;
+    }
+    /// @notice Returns the length of ApkUpdates for the provided `quorumNumber`
+    function getQuorumApkHistoryLength(uint8 quorumNumber) external view returns(uint32){
+        return uint32(quorumApkUpdates[quorumNumber].length);
+    }
+    /// @notice Returns the length of ApkUpdates for the global APK
+    function getGlobalApkHistoryLength() external view returns(uint32){
+        return uint32(globalApkUpdates.length);
+    }
+
+    function _processGlobalApkUpdate(BN254.G1Point memory point) internal {
+        // load and store in memory in common case we need to access the length again
+        uint256 globalApkUpdatesLength = globalApkUpdates.length;
+        // update the nextUpdateBlockNumber of the previous update
+        if (globalApkUpdates.length > 0) {
```

Comment:
immediately not using the thing we just stored in memory, and instead SLOADing again

Reply:
whoops